### DISCUSSION TITLE
FaceDivFree interpolation for a staggered field

### DISCUSSION THREAD

**Discussion by farhaddaei** [AMReX-Codes/amrex#4075] [Q&A] [answered]
Hi everyone!

In our code, we have a staggered magnetic field (each component is **_face-centred_** in its respective direction), so each component has its own index type. Therefore, in the **variableSetUp** method of the class inheriting from the AmrLevel class, we need to add a separate element to **desc_lst** (correct?). However, to use the **face_divfree_interp** interpolator (of type **FaceDivFree**, to preserve the divergence of the magnetic field) we need to inform AMReX that the components provided in separate elements of **desc_lst** should participate in the interpolation. Unfortunately, I did not find any example of such interpolation in the tutorials, code examples, some other codes which use AMReX, or any explicit instruction about this subject in the AMReX documentation.

I appreciated any hint about using the mentioned interpolation method.

Regards,
Farhad

**Comment by WeiqunZhang** [comment, marked as answer]
I assume you wanted to use the divergence free interpolater in FillPath. You should be able to call https://github.com/AMReX-Codes/amrex/blob/e792933f9561c9ab0e47238c041a6c09818f0908/Src/AmrCore/AMReX_FillPatchUtil.H#L277 and https://github.com/AMReX-Codes/amrex/blob/e792933f9561c9ab0e47238c041a6c09818f0908/Src/AmrCore/AMReX_FillPatchUtil.H#L685 directly without using `AmrLevel::FillPatch`. For convenience, you might want to create a function say `FillPatchBfield` that does what `AmrLevel::FillPatch` does for other variables.

Aslo, `AmrLevel` has two virtual functions `virtual void init (AmrLevel &old)` and `virtual void init ()` that are used during regrid. Here is an example. https://github.com/AMReX-Fluids/IAMR/blob/37327b3ed63d3ac29e2c6f2937aa071a2b06e02c/Source/NavierStokesBase.cpp#L1713. In these two functions, you should be able to call `FillPatchTwoLevels` and `InterpFromCoarseLevel` directly or your `FillPatchBfield`  instead of `FillPatch` for the B field.

**Reply by farhaddaei** [reply]
@WeiqunZhang thanks for the hints and the link to IAMR repo. As you mentioned we needed **face_divfree_interp** to fill patches in both directions (coarse <=> fine) and the methods you suggested are perfect for the task.
We will pack magnetic field components as implemented in ["amrex/Tests/DivFreePatch/](https://github.com/AMReX-Codes/amrex/blob/e792933f9561c9ab0e47238c041a6c09818f0908/Tests/DivFreePatch/main.cpp#L354)"

Just out of curiosity, how could one use the "FaceDivFree" class?  The name and the description of the class (the comments in the source code) are very tempting for scenarios including "AMR + staggered magnetic field"

**Reply by WeiqunZhang** [reply]
One uses ` FaceDivFree` when calling fillpatch functions by passing `&amrex::face_divfree_interp` as the `Interp*` argument. For MHD, you probably also need `amrex::EdgeFluxRegister` for E (or v x B).
